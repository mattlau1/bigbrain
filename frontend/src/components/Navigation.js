import React from 'react';
import Navbar from 'react-bootstrap/Navbar';
import Nav from 'react-bootstrap/Nav';
import styled from 'styled-components';
import { Link, NavLink, useHistory } from 'react-router-dom';
import API from '../utils/API';
import { useAlert } from '../contexts/AlertProvider';
import Logo from '../assets/Logo.png';
import { Image } from 'react-bootstrap';

const Brand = styled.span`
  font-size: 20pt;
`

const Navigation = () => {
  const history = useHistory();
  const dispatch = useAlert();
  const api = new API();

  const createAlert = (type, message) => {
    dispatch({
      type: type,
      message: message,
    })
  }

  const logOut = async () => {
    try {
      // we want users to be logged out on the client side
      // even if the server side request fails
      const token = localStorage.getItem('token');
      localStorage.removeItem('token');
      history.push('/');
      createAlert('SUCCESS', 'Successfully logged out')

      // server side log out
      const res = await api.postAPIRequestToken('admin/auth/logout', token)
      const data = await res.json();
      if (!res.ok) {
        console.warn(data.message);
      }
    } catch (e) {
      console.warn(e);
    }
  }

  return (
    <>
      <Navbar collapseOnSelect expand="lg" bg="dark" variant="dark">
        <Navbar.Brand as={Link} to="/dashboard" className="d-flex align-items-center mx-2">
          <Image
            style={{
              width: 60,
              height: 60,
            }}
            src={Logo}
            className="d-inline-block align-top mr-2"
          />
          <Brand id="brand-name">BigBrain</Brand>
        </Navbar.Brand>

        <Navbar.Toggle aria-controls="responsive-navbar-nav" id='navButton'/>
        <Navbar.Collapse id="responsive-navbar-nav">
          <Nav variant="pills" className="ml-auto" defaultActiveKey="/creategame">
            <Nav.Link as={NavLink} to="/dashboard" id="dashboard-link" className="mx-2 text-center">
              Dashboard
            </Nav.Link>
            <Nav.Link
              as={NavLink}
              className="mx-2 text-center"
              exact
              to="/"
              onClick={() => {
                logOut();
              }}
              id='signout-link'
            >
              Sign Out
            </Nav.Link>
          </Nav>
        </Navbar.Collapse>
      </Navbar>
    </>
  );
}

export default Navigation
