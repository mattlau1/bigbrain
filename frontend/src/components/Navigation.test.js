import { shallow } from 'enzyme';
import React from 'react';
import Navigation from './Navigation';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';

describe('Navigation', () => {
  it('should display BigBrain as the brand', () => {
    const tree = renderer.create(<Router><Navigation /></Router>).toJSON();
    const wrapper = shallow(<Navigation />);
    expect(wrapper).toHaveLength(1);
    const text = wrapper.find('#brand-name');
    expect(text).toHaveLength(1);
    expect(text.text()).toBe('BigBrain');
    expect(tree).toMatchSnapshot();
  })

  it('should not have a collapsed navbar button', () => {
    const tree = renderer.create(<Router><Navigation /></Router>).toJSON();
    const wrapper = shallow(<Navigation />);
    const navbar = wrapper.find('#nav-button');
    expect(navbar.hasClass('navbar-toggler collapsed')).toBeFalsy();
    expect(wrapper.find('.show')).toHaveLength(0);
    expect(tree).toMatchSnapshot();
  })

  it('should have dashboard button', () => {
    const tree = renderer.create(<Router><Navigation /></Router>).toJSON();
    const wrapper = shallow(<Navigation />);
    const text = wrapper.find('#dashboard-btn');
    expect(text.text()).toBe('Dashboard');
    expect(tree).toMatchSnapshot();
  })

  it('should have signout button', () => {
    const tree = renderer.create(<Router><Navigation /></Router>).toJSON();
    const wrapper = shallow(<Navigation />);
    const text = wrapper.find('#logout-btn');
    expect(text.text()).toBe('Sign Out');
    expect(tree).toMatchSnapshot();
  })
})
