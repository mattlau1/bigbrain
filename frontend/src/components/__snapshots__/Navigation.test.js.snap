// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation should display BigBrain as the brand 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <a
    className="d-flex align-items-center mx-2 navbar-brand"
    href="/dashboard"
    onClick={[Function]}
  >
    <img
      className="d-inline-block align-top mr-2"
      src="logo.png"
      style={
        Object {
          "height": 60,
          "width": 60,
        }
      }
    />
    <span
      className="sc-bdvvaa kgoOxA"
      id="brand-name"
    >
      BigBrain
    </span>
  </a>
  <button
    aria-controls="responsive-navbar-nav"
    aria-label="Toggle navigation"
    className="navbar-toggler collapsed"
    id="nav-button"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    aria-expanded={null}
    className="navbar-collapse collapse"
    id="responsive-navbar-nav"
  >
    <div
      className="ml-auto navbar-nav nav-pills"
      onKeyDown={[Function]}
    >
      <a
        aria-current={null}
        className="mx-2 text-center nav-link"
        data-rb-event-key={null}
        disabled={false}
        href="/dashboard"
        id="dashboard-btn"
        onClick={[Function]}
      >
        Dashboard
      </a>
      <a
        aria-current="page"
        className="mx-2 text-center nav-link active"
        data-rb-event-key={null}
        disabled={false}
        href="/"
        id="logout-btn"
        onClick={[Function]}
        style={Object {}}
      >
        Sign Out
      </a>
    </div>
  </div>
</nav>
`;

exports[`Navigation should have dashboard button 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <a
    className="d-flex align-items-center mx-2 navbar-brand"
    href="/dashboard"
    onClick={[Function]}
  >
    <img
      className="d-inline-block align-top mr-2"
      src="logo.png"
      style={
        Object {
          "height": 60,
          "width": 60,
        }
      }
    />
    <span
      className="sc-bdvvaa kgoOxA"
      id="brand-name"
    >
      BigBrain
    </span>
  </a>
  <button
    aria-controls="responsive-navbar-nav"
    aria-label="Toggle navigation"
    className="navbar-toggler collapsed"
    id="nav-button"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    aria-expanded={null}
    className="navbar-collapse collapse"
    id="responsive-navbar-nav"
  >
    <div
      className="ml-auto navbar-nav nav-pills"
      onKeyDown={[Function]}
    >
      <a
        aria-current={null}
        className="mx-2 text-center nav-link"
        data-rb-event-key={null}
        disabled={false}
        href="/dashboard"
        id="dashboard-btn"
        onClick={[Function]}
      >
        Dashboard
      </a>
      <a
        aria-current="page"
        className="mx-2 text-center nav-link active"
        data-rb-event-key={null}
        disabled={false}
        href="/"
        id="logout-btn"
        onClick={[Function]}
        style={Object {}}
      >
        Sign Out
      </a>
    </div>
  </div>
</nav>
`;

exports[`Navigation should have signout button 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <a
    className="d-flex align-items-center mx-2 navbar-brand"
    href="/dashboard"
    onClick={[Function]}
  >
    <img
      className="d-inline-block align-top mr-2"
      src="logo.png"
      style={
        Object {
          "height": 60,
          "width": 60,
        }
      }
    />
    <span
      className="sc-bdvvaa kgoOxA"
      id="brand-name"
    >
      BigBrain
    </span>
  </a>
  <button
    aria-controls="responsive-navbar-nav"
    aria-label="Toggle navigation"
    className="navbar-toggler collapsed"
    id="nav-button"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    aria-expanded={null}
    className="navbar-collapse collapse"
    id="responsive-navbar-nav"
  >
    <div
      className="ml-auto navbar-nav nav-pills"
      onKeyDown={[Function]}
    >
      <a
        aria-current={null}
        className="mx-2 text-center nav-link"
        data-rb-event-key={null}
        disabled={false}
        href="/dashboard"
        id="dashboard-btn"
        onClick={[Function]}
      >
        Dashboard
      </a>
      <a
        aria-current="page"
        className="mx-2 text-center nav-link active"
        data-rb-event-key={null}
        disabled={false}
        href="/"
        id="logout-btn"
        onClick={[Function]}
        style={Object {}}
      >
        Sign Out
      </a>
    </div>
  </div>
</nav>
`;

exports[`Navigation should not have a collapsed navbar button 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <a
    className="d-flex align-items-center mx-2 navbar-brand"
    href="/dashboard"
    onClick={[Function]}
  >
    <img
      className="d-inline-block align-top mr-2"
      src="logo.png"
      style={
        Object {
          "height": 60,
          "width": 60,
        }
      }
    />
    <span
      className="sc-bdvvaa kgoOxA"
      id="brand-name"
    >
      BigBrain
    </span>
  </a>
  <button
    aria-controls="responsive-navbar-nav"
    aria-label="Toggle navigation"
    className="navbar-toggler collapsed"
    id="nav-button"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    aria-expanded={null}
    className="navbar-collapse collapse"
    id="responsive-navbar-nav"
  >
    <div
      className="ml-auto navbar-nav nav-pills"
      onKeyDown={[Function]}
    >
      <a
        aria-current={null}
        className="mx-2 text-center nav-link"
        data-rb-event-key={null}
        disabled={false}
        href="/dashboard"
        id="dashboard-btn"
        onClick={[Function]}
      >
        Dashboard
      </a>
      <a
        aria-current="page"
        className="mx-2 text-center nav-link active"
        data-rb-event-key={null}
        disabled={false}
        href="/"
        id="logout-btn"
        onClick={[Function]}
        style={Object {}}
      >
        Sign Out
      </a>
    </div>
  </div>
</nav>
`;
